src/value.rs

```diff
@@ -0,0 +1,298 @@
+
+use std::{borrow::Cow, collections::BTreeMap};
+
+#[cfg(feature = "serde")]
+use std::{
+    convert::TryInto,
+    fmt::{self, Formatter},
+};
+
+#[cfg(feature = "serde")]
+use serde_bytes::Bytes;
+
+#[cfg(feature = "serde")]
+use serde::{
+    ser::{SerializeMap, SerializeSeq},
+    Serialize,
+};
+
+use crate::{
+    decoding::{FromBencode, Object},
+    encoding::{SingleItemEncoder, ToBencode},
+};
+
+/// An owned or borrowed bencoded value.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum Value<'a> {
+    /// An owned or borrowed byte string
+    Bytes(Cow<'a, [u8]>),
+    /// A dictionary mapping byte strings to values
+    Dict(BTreeMap<Cow<'a, [u8]>, Value<'a>>),
+    /// A signed integer
+    Integer(i64),
+    /// A list of values
+    List(Vec<Value<'a>>),
+}
+
+impl<'a> Value<'a> {
+    /// Convert this Value into an owned Value with static lifetime
+    pub fn into_owned(self) -> Value<'static> {
+        match self {
+            Value::Bytes(bytes) => Value::Bytes(Cow::Owned(bytes.into_owned())),
+            Value::Dict(dict) => Value::Dict(
+                dict.into_iter()
+                    .map(|(key, value)| (Cow::Owned(key.into_owned()), value.into_owned()))
+                    .collect(),
+            ),
+            Value::Integer(integer) => Value::Integer(integer),
+            Value::List(list) => Value::List(list.into_iter().map(Value::into_owned).collect()),
+        }
+    }
+}
+
+impl<'a> ToBencode for Value<'a> {
+    const MAX_DEPTH: usize = usize::max_value();
+
+    fn encode(&self, encoder: SingleItemEncoder) -> Result<(), crate::encoding::Error> {
+        match self {
+            Self::Bytes(bytes) => encoder.emit_bytes(bytes),
+            Self::Dict(dict) => dict.encode(encoder),
+            Self::Integer(integer) => integer.encode(encoder),
+            Self::List(list) => list.encode(encoder),
+        }
+    }
+}
+
+impl<'a> FromBencode for Value<'a> {
+    fn decode_bencode_object(object: Object) -> Result<Self, crate::decoding::Error> {
+        match object {
+            Object::Bytes(bytes) => Ok(Value::Bytes(Cow::Owned(bytes.to_owned()))),
+            Object::Dict(mut decoder) => {
+                let mut dict = BTreeMap::new();
+                while let Some((key, value)) = decoder.next_pair()? {
+                    dict.insert(
+                        Cow::Owned(key.to_owned()),
+                        Value::decode_bencode_object(value)?,
+                    );
+                }
+                Ok(Value::Dict(dict))
+            },
+            Object::Integer(text) => Ok(Value::Integer(text.parse()?)),
+            Object::List(mut decoder) => {
+                let mut list = Vec::new();
+                while let Some(object) = decoder.next_object()? {
+                    list.push(Value::decode_bencode_object(object)?);
+                }
+                Ok(Value::List(list))
+            },
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
```

Comment:
I am totally fine with this, but a pattern that I have found that makes optional impls much nicer to read and write is
```rust
#[cfg(feature = "serde")]
mod serde_impls {
   use super::*;
   struct HelperType;
   impl Serialize for Value {}
   impl Deserialize for Value {}
}
```

TBH, there's not much reason to use one over the other, but I thought that I should point it out in case you were unaware of the pattern.

Reply:
Nice, that's much better.